spec/models/provider_spec.rb

```diff
@@ -0,0 +1,73 @@
+require 'rails_helper'
+
+RSpec.describe Provider, type: :model do
+
+  describe "Provider" do
+    it "is provider valid" do
+      expect(build(:provider)).to be_valid
+    end
+  end
+
+  describe "Validations" do
+    it { is_expected.to validate_uniqueness_of(:name) }
+  end
+
+  describe "Associations" do
+    it "has many warehouse_items" do
+      assc = described_class.reflect_on_association(:warehouse_items)
+      expect(assc.macro).to eq :has_many
+    end
+
+    it "has many warehouse_item_providers" do
+      assc = described_class.reflect_on_association(:warehouse_item_providers)
+      expect(assc.macro).to eq :has_many
+    end
+  end
+
+  describe "before_save" do
+     it { is_expected.to callback(:set_name).before(:save) }
```

Comment:
Te falta probar las situaciones donde se asigna el nombre y se prueban los resultados. Este it solo esta probando que corra el callback. Tambien hay que probar los resultados y todas las posibles vartiantes